### ISSUE TITLE
On macOS, crashes when a system tray is created before window initialization.

### ISSUE THREAD

**Issue by tasuren** [tauri-apps/tao#774]
I am not familiar with English and I apologize if I have been rude.

**Describe the bug**
Using the system tray feature under macOS and trying to build a system tray with `SystemTrayBuilder` at a time before the window is initialized causes a crash.

This problem is caused by `SystemTrayBuilder::new` setting up `NSStatusItem` at [here](https://github.com/tauri-apps/tao/blob/75eb0c1e7e83a766af0e083ce09c761d1974cde4/src/platform_impl/macos/system_tray.rs#L43) before `NSApplication` is initialized. And the reason why initializing a window before using a system tray fixes this problem is probably tao's window struct will use `NSApplication` at [here](https://github.com/tauri-apps/tao/blob/75eb0c1e7e83a766af0e083ce09c761d1974cde4/src/platform_impl/macos/window.rs#L313-L352) or other place.

Also, if you run the system tray example with the tao at the time of commit `972307ddf088b0f941be2ea66bded2473222aed5`, this problem does not happen. I think this is because `EventLoop::new` uses `NSApplication` first at [here](https://github.com/tauri-apps/tao/blob/972307ddf088b0f941be2ea66bded2473222aed5/src/platform_impl/macos/event_loop.rs#L126C35-L126C35). And `EventLoop::new` of the latest code doesn't use `NSApplication`, so the system tray example will fail with the latest tao.

So this problem is quickly fixed by running the following function before setting up a system tray to set up `NSApplication` first.
```rust
use objc::{*, runtime::Object};

fn setup_application() {
  unsafe {
    let _: *const Object = msg_send![class!(NSApplication), sharedApplication];
  };
}
```

**Steps To Reproduce**
If you want to reproduce crush, you can do it by following these steps:
1. Clone this repository.
2. Run the system tray example by: `cargo run --features tray --example system_tray`
3. It will crush with a message like below.
  `[1]    6367 trace trap  cargo run --features tray --example system_tray`
  or `Assertion failed: (CGAtomicGet(&is_initialized)), function CGSConnectionByID, file CGSConnection.mm, line 422.`

Or no crush steps:
1. Clone this repository.
2. Insert the following code just below [the part that creates the event loop](https://github.com/tauri-apps/tao/blob/75eb0c1e7e83a766af0e083ce09c761d1974cde4/examples/system_tray.rs#L24C1-L24C1).
  ```rust
  let window = tao::window::WindowBuilder::new().build(&event_loop).unwrap();
  ```
3. Run the system tray example: `cargo run --features tray --example system_tray`
4. It works but with window.

**Expected behavior**
No crush.

**Screenshots**
None

**Platform and Versions (please complete the following information):**
OS: macOS 13.4.1 22F82 arm64
Rustc: 1.71.0 (8ede3aae2 2023-07-12)

**Would you want to assign yourself to resolve this bug?**
I still don't know much about macOS API's and I don't know much about tao.
~~So I think I can just assign the code to get `NSApplication` from `APP_CLASS` in `src/platform_impl/macos/app.rs` or `cocoa::appkit::NSApp` to the constructor of `SystemTrayBuilder`, but~~^[1] I may really have to do something more complicated.
And I am busy right now.
So if the changes are small and not too time consuming, I think I will use the strength to fix this bug.

^[1]: I was looking at the code and changed my mind. See my next message for details.

**Additional context**
By the way, it works fine on the latest tauri. This probably has something to do with the window initialization timing I mentioned earlier or something.

**Comment by tasuren** [comment]
It appears that the problem occurs in tao after [this commit](https://github.com/tauri-apps/tao/commit/093d8fbc20954d51c96c94618479ff465ca55888#diff-04e18fafd5bcabb8761ec52f10341b4b0c55c784aa1b8af81c77e732b7b8ec3b).
This commit moves the code that sets up `NSApplication` from `EventLoop::new` to `run_return` method in `EventLoop`. This is the source of the problem because, as I said earlier, it delays the initialization of the `NSApplication` and causes the `examples/system_tray.rs` execution to crash.

Also, I may have found another problem that I can't do [the methods of EventLoopExtMacOS](https://github.com/tauri-apps/tao/blob/75eb0c1e7e83a766af0e083ce09c761d1974cde4/src/platform/macos.rs#L485-L514) before starting the event loop. This problem also occurs after the above commit.
I think this problem is caused by `EventLoop`'s `delegate` is `nil` until `run_return` of `EventLoop` runs. And at the time before the above commit, the delegate is not `nil` and this problem is not show up because the delegate is initialized at `EventLoop::new`.

In summary, these two problems are made by the above commit that moves the initialization timing for `NSApplication` and `delegate` of `EventLoop` from `EventLoop::new` to `EventLoop::run_return`.
This problem is caused by the same part of this issue. Should I start an Issue as a separate Issue from this Issue? Or should I abstract the two issues and edit the title of this Issue as one Issue?

For my part, I think that delaying the initialization timing from `new` to `run_return` is to make the event loop and `NSApplication` run in the main thread and to set the main thread to the place where the initialization is executed, so the decision of the above commit is correct and reverting to `new` so easily is I don't think it is a good idea.
And one thing I just thought of is to have a mechanism to generate a panic or compile error if the thread at the time of `new` runs and the thread at the time of `run_return` runs are different. In such a case, I think initialization with `new` is not likely to cause any problems.
In other words, if you see the two aforementioned problems as being about initialization timing, I thought I could edit the title of this Issue, etc. to make the two Issues into one Issue about initialization timing.

**Comment by amrbashir** [comment]
cc @lucasfernog

**Comment by lucasfernog** [comment]
I just faced this today in tauri too so I've opened a PR to fix it. Thanks!
